### PR TITLE
feat(core): load genesis file by network id

### PR DESCRIPTION
**Pull request by dan-da**
Addresses #243

Support for loading a different genesis block for each network.

The genesis block is loaded from a file whose path is the pattern: `~/.kindelia/genesis/<network-id>.kdl`

Changes:

 core:
  * add pub(crate) util::genesis_path() with associated error enum
  * add pub(crate) util::genesis_code() with associated error enum
  * modify hvm::test_statements*() to accept network_id and load genesis block from file instead of compiled string
  * modify node::new() to to accept network_id and load genesis block from file instead of compiled string

note: the error enums are based on the style/pattern recommended in [this article](https://mmapped.blog/posts/12-rust-error-handling.html), which is well worth a read.  The pre-existing calling functions still call unwrap() on the result, but intent is to address that in #247.

 cli:
  * add network_id to test command, required by hvm::test_statements()
  * fix parsing of hex values for --network_id
  * add clap_num dep for parsing hex values
 
---------

I will note that this change might have security implications.  In particular, when a genesis block is loaded from a static string in the executable then a checksum of the executable binary will also include the genesis block.  By moving the block out into the filesystem we lose that guarantee.  So both the binary and corresponding block file must be summed independently.    Of course the benefit is that the system can cleanly handle an arbitrary number of networks.  It could be that for mainnet, we use a hybrid approach, whereby mainnet and perhaps present testnet(s) genesis blocks are included in the binary, but others can be created by users at will.

### CODE REVIEW

**Comment by dan-da** [comment]
I think tests are failing in CI because the genesis file does not exist yet.  They pass on my machine, where it does exist.  Probably I need to update the `kindelia init` command to create this file.  I thought about this a couple times and then forgot about it.  I will look at it tmw.

File: kindelia/src/cli.rs [hunk 1]
```diff
@@ -120,6 +121,9 @@ pub enum CliCommand {
     /// Whether to consider size and mana in the execution.
     #[clap(long)]
     sudo: bool,
+    /// Network id / magic number.
+    #[clap(long, value_parser=maybe_hex::<u32>)]
+    network_id: Option<u32>,
```
Comment (steinerkelvin) [inline kindelia/src/cli.rs:126]: Maybe it makes more sense to accept the "prelude" code itself here instead of the network id.
"What does offline testing has to do with the network id?"

File: kindelia_core/src/hvm.rs [hunk 2]
```diff
@@ -4245,7 +4244,7 @@ pub fn print_io_consts() {
 }
 
 // Serializes, deserializes and evaluates statements
-pub fn test_statements(statements: &Vec<Statement>, debug: bool) {
+pub fn test_statements(network_id: u32, statements: &Vec<Statement>, debug: bool) {
```
Comment (steinerkelvin) [inline kindelia_core/src/hvm.rs:4247]: Maybe here applies the same comment I made on `Node::new`, hmm

File: kindelia_core/src/node.rs [hunk 3]
```diff
@@ -774,7 +773,7 @@ impl<C: ProtoComm, S: BlockStorage> Node<C, S> {
     let (query_sender, query_receiver) = mpsc::sync_channel(1);
 
     let genesis_stmts =
-      parser::parse_code(constants::GENESIS_CODE).expect("Genesis code parses");
+      parser::parse_code(&genesis_code(network_id).unwrap()).expect("Genesis code parses");
```
Comment (steinerkelvin) [inline kindelia_core/src/node.rs:776]: This should not implicitly load another file (or even do IO at all) but receive the source code (or the already parsed statements from outside).

`Node` implements a core logic that we are decoupling the most possible from IO, so we can mock on tests, test/benchmark multiple implementations etc. Take a look at `C: ProtoComm` and `S: BlockStorage` for example.

File: kindelia_core/src/hvm.rs [hunk 4]
```diff
@@ -4256,7 +4255,7 @@ pub fn test_statements(statements: &Vec<Statement>, debug: bool) {
 
   // TODO: code below does not need heaps_path at all. extract heap persistence out of Runtime.
   let heaps_path = dirs::home_dir().unwrap().join(".kindelia").join("state").join("heaps");
-  let genesis_smts = parse_code(constants::GENESIS_CODE).expect("Genesis code parses");
+  let genesis_smts = parse_code(&genesis_code(network_id).unwrap()).expect("Genesis code parses");
```
Comment (steinerkelvin) [inline kindelia_core/src/hvm.rs:4258]: See comment on `Node::new`.

File: kindelia/src/cli.rs [hunk 5]
```diff
@@ -3,6 +3,7 @@ use std::path::PathBuf;
 
 use clap::{Parser, Subcommand};
 use clap_complete::Shell;
+use clap_num::maybe_hex;
```
Comment (steinerkelvin) [inline kindelia/src/cli.rs:6]: This is nice :v

**Review by steinerkelvin — CHANGES_REQUESTED** [review]

**Comment by steinerkelvin** [comment]
> update the kindelia init command to create this file
Totally. It should be done inside (or beside) `init_config_file()`.

File: kindelia/src/cli.rs (hunk 1 above)
Comment (dan-da) [reply kindelia/src/cli.rs:126]: > Maybe it makes more sense to accept the "prelude" code itself here instead of the network id.

So accept the code itself as a CLI argument?   And in the config file?  That's a potentially very long and multiline string...   I don't see it.

> What does offline testing has to do with the network id?

I wondered this myself.   But this code eventually calls `hvm::test_statements()` which has a call `parse_code(GENESIS_CODE)`.   And `GENESIS_CODE` is defined as `include_str!("../genesis.kdl")`. 

In other words, this `check` command is relying on GENESIS_CODE exactly as Node::new() does.  And GENESIS_CODE is the thing we are trying to make dynamic per network, as I understand it.

I did not look into it deeply, but rather assumed that the tests may rely on the prelude/genesis statements.

So I guess I would ask you the same question.  Why is this needed?

File: kindelia_core/src/hvm.rs (hunk 2 above)
Comment (dan-da) [reply kindelia_core/src/hvm.rs:4247]: ok, I will reply to that one.

File: kindelia_core/src/node.rs (hunk 3 above)
Comment (dan-da) [reply kindelia_core/src/node.rs:776]: yeah that's fine.  I actually started out doing it that way but it was going to require more changes to callers, so I went with the approach that changed fewest LOC.

So the basic idea then, as I see it, is that `kindelia start` command receives the network_id and the cli layer reads the corresponding `~/kindelia/genesis/<network-id>.kdl file`, parses the statements, and then passes `Vec<Statement>` to `Node::new()`.   Sound good?

File: kindelia_core/src/hvm.rs (hunk 4 above)
Comment (dan-da) [reply kindelia_core/src/hvm.rs:4258]: replied to Node::new() comment.

File: kindelia/src/cli.rs (hunk 5 above)
Comment (dan-da) [reply kindelia/src/cli.rs:6]: yeah, I was happy to find it.    nicer if it were part of core Clap and didn't require extra dep though.

File: kindelia/src/cli.rs (hunk 1 above)
Comment (dan-da) [reply kindelia/src/cli.rs:126]: ok, I see why it is using the genesis code.  The prelude is setting up required basics.  

I found that `test_statements()` is passing the parsed code to `init_runtime()`.   I tried passing an empty vec instead.  Then I ran `kindelia test example/example.kdl`:

```
$ kindelia test ../example/example.kdl 
Block 
=====

0000000000 [run] ERROR: 'Call' is not defined.
```

So it seems to be truly necessary.  I will get rid of `network_id` and pass `&[Statement]`.

**Comment by dan-da** [comment]
Ok, so I just pushed a commit that does the `kindelia init` genesis install and also passes `&[Statement]` to core api instead of network_id.

Getting the init functionality working was a bit tricky.   I had to think about how these genesis files will be stored in the git repo and installed.  I figured it makes sense that they would have same filename in the repo as in user's homedir.   I put them in `kindelia` crate since all the loading is taking place there now.

Previously the genesis code was loaded into the executable with `include_str!("genesis.kdl")` macro, which works fine.  But it requires a static string for the file path.  Now I wanted the code to dynamically include the most recent version, eg `"0xCAFE0006.kdl"`.  That should be possible with a macro, but I am not experienced writing rust macros yet.  So I found `include_dir!()` crate/macro that includes all files in a given directory.    With this, I am able to sort and get the latest version for installation, which works.

However that gets me thinking:  with this `include_dir!()` functionality including multiple genesis files directly in the source code, do we really need/want to load genesis files from disk at runtime?     For purposes of testnets and mainnet, would it not be sufficient (and less complicated) to have a few directly in the binary?   (and no need to install + load)

The tradeoff is that users would not be able to define their own networks without recompiling.  I am unsure that users would typically *need* to define their own networks though...?

I hope I am being clear.    ;-)     See also the kindelia/genesis/README.md

**Comment by dan-da** [comment]
rebased to dev tip.

File: kindelia_core/src/node.rs (hunk 3 above)
Comment (dan-da) [reply kindelia_core/src/node.rs]: Node::new now receives &[Statement], so marking this resolved.

File: kindelia/src/cli.rs (hunk 1 above)
Comment (dan-da) [reply kindelia/src/cli.rs:126]: not marking this resolved yet, because as I see it, `--network-id` needs to be an argument for `kindelia check` because the test statements typically depend on the prelude, which varies by network.

@steinerkelvin do you agree?

**Comment by dan-da** [comment]
I don't understand why the tests are failing in CI.   The errors indicate that `~/.kindelia/genesis/0xCAFE0006.kdl` is not found.

But it works fine for me, provided that I run `kindelia init`.   Example:

```
$ cargo install --path .
...
$ mv ~/.kindelia/ ~/.kindelia.bak

$ kindelia init
Writing default configuration to '/home/danda/.kindelia/kindelia.toml'...

$ tree ~/.kindelia/
/home/danda/.kindelia/
├── genesis
│   └── 0xCAFE0006.kdl
└── kindelia.toml

1 directory, 2 files

$ cargo test
...
test result: ok. 37 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 8.44s
```

So either the CI is somehow not running `kindelia init` or there is something else different in the environment I do not understand.

File: kindelia/src/cli.rs (hunk 1 above)
Comment (steinerkelvin) [reply kindelia/src/cli.rs:126]: > So accept the code itself as a CLI argument?

A file containing the code. If not provided it would use the `include_str!("../genesis.kdl")` on the binary (see my comment).

> --network-id needs to be an argument for kindelia check

I agree! `kindelia check` will be a typically online command (it would check if names are taken etc), it needs to talks with a node and so is aware of the network id.
But `kindelia test` should be offline, so if it takes at most a parameter that would be something like a `--prelude` one, that would point to a file with alternative initial statements. The default would be to use the `genesis.kdl` included in the binary — we should avoid reading files as possible.

**Comment by steinerkelvin** [comment]
> I am unsure that users would typically need to define their own networks though

I think this is useful for development / experimentation. But it's indeed not strictly necessary.

> With this, I am able to sort and get the latest version for installation, which works.

I was thinking of keeping `include_str!("genesis.kdl")`, hardcoding the "default" besides it (in this version `0xCAFE0006`) that would be filled in the `default.toml` template to be written on the user config, and would be the filename for `genesis.kdl` when installed.

It's important to keep `genesis.kdl` static - somehow - because it's used in some tests. These unit tests should not depend on external files at all.

> So either the CI is somehow not running kindelia init or there is something else different in the environment I do not understand.

It's indeed not run, this should not be necessary to run tests.

**Comment by dan-da** [comment]
> It's important to keep genesis.kdl static - somehow - because it's used in some tests. These unit tests should not depend on external files at all.

> It's indeed not run, this should not be necessary to run tests.

I thought I had accounted for this by creating genesis-tests.kdl which the tests include_str!().   

What I missed is that some tests actually invoke the kindelia binary with a kindelia!() macro and then check its output.  And since I had already run `kindelia init` locally, the test succeeded.

I think what I can do is change the `test` command to accept a file path as you suggest.  Then the tests can pass that file to the binary.

Anyway, I will be sure and run the tests without the ~/.kindelia directory existing before pushing to this branch again.

File: kindelia/src/cli.rs (hunk 1 above)
Comment (dan-da) [reply kindelia/src/cli.rs]: I modified `kindelia test` to accept a `--genesis` arg that allows user to provide a path to  prelude file, or else it defaults to `kindelia_core/genesis-tests.kdl` which is compiled into the executable.

**Comment by dan-da** [comment]
I went back to the drawing board a little on this, thinking about what is useful and makes sense for the end user.

Overview:
* same: kindelia/genesis/networks/*.kdl are compiled into binary and considered built-in networks.  Any number of built-in networks can be supported simultaneously.
* same: user can choose between built-in networks using `network_id` (env, conf, cli arg)
* new: user can specify a `network_id` that is *not built-in* and also specify a `genesis` file containing prelude statements (env, conf, cli arg)
* new: if user attempts to specify a `genesis` file for built-in network, an error is printed and node does not start.  The idea here is to prevent confusion/weirdness if a user mistakenly (or knowingly) attempts to run non standard prelude with a built-in network, eg mainnet.
* fixed: `test` command accepts a `genesis` file from user, or else defaults to `kindelia_core/genesis-tests.kdl`
* fixed: kindelia_core/genesis-tests.kdl is compiled into binary and used for test cases in both kindelia_core and kindelia.

**Comment by dan-da** [comment]
rebased on latest dev.

**Comment by dan-da** [comment]
rebased on latest dev to resolve conflict.

**Comment by dan-da** [comment]
rebased on latest dev to resolve conflict.
